src/main/java/org/testng/internal/Invoker.java

```diff
@@ -474,109 +387,101 @@ else if (! m_continueOnFailedConfiguration) {
 
-        //
-        // If this method is a IConfigurable, invoke its run() method
-        //
-        IConfigurable configurableInstance =
-          IConfigurable.class.isAssignableFrom(method.getDeclaringClass()) ?
-          (IConfigurable) targetInstance : m_configuration.getConfigurable();
-        if (configurableInstance != null) {
-          MethodInvocationHelper.invokeConfigurable(targetInstance, params, configurableInstance, method.getMethod(),
-              testResult);
-        }
-        else {
-          //
-          // Not a IConfigurable, invoke directly
-          //
-          if (MethodHelper.calculateTimeOut(tm) <= 0) {
-            MethodInvocationHelper.invokeMethod(method.getMethod(), targetInstance, params);
-          }
-          else {
-            MethodInvocationHelper.invokeWithTimeout(tm, targetInstance, params, testResult);
-            if (!testResult.isSuccess()) {
-              // A time out happened
-              throwConfigurationFailure(testResult, testResult.getThrowable());
-              throw testResult.getThrowable();
-            }
-          }
-        }
-      }
-      catch (InvocationTargetException | IllegalAccessException ex) {
-       throwConfigurationFailure(testResult, ex);
-       throw ex;
-      } catch (Throwable ex) {
-        throwConfigurationFailure(testResult, ex);
-        throw new TestNGException(ex);
-      }
-      finally {
-        testResult.setEndMillis(System.currentTimeMillis());
-        Reporter.setCurrentTestResult(testResult);
-        runInvokedMethodListeners(AFTER_INVOCATION, invokedMethod, testResult);
-        Reporter.setCurrentTestResult(null);
+      IConfigurable configurableInstance = computeConfigurableInstance(method, targetInstance);
+      if (configurableInstance != null) {
+        MethodInvocationHelper.invokeConfigurable(targetInstance, params,
+                configurableInstance, method.getMethod(), testResult);
+      } else {
+        MethodInvocationHelper.invokeMethodConsideringTimeout(tm, method, targetInstance, params, testResult);
       }
+    } catch (InvocationTargetException | IllegalAccessException ex) {
+      throwConfigurationFailure(testResult, ex);
+      throw ex;
+    } catch (Throwable ex) {
+      throwConfigurationFailure(testResult, ex);
+      throw new TestNGException(ex);
+    } finally {
+      testResult.setEndMillis(System.currentTimeMillis());
+      Reporter.setCurrentTestResult(testResult);
+      runInvokedMethodListeners(AFTER_INVOCATION, invokedMethod, testResult);
+      Reporter.setCurrentTestResult(null);
     }
   }
 
-  private void throwConfigurationFailure(ITestResult testResult, Throwable ex)
-  {
+  private IConfigurable computeConfigurableInstance(ConstructorOrMethod method, Object targetInstance) {
+    return IConfigurable.class.isAssignableFrom(method.getDeclaringClass()) ?
+            (IConfigurable) targetInstance : m_configuration.getConfigurable();
```

Comment:
I would refrain from making those changes as part of this PR because it kind of deviates the intent. The intent of this PR for me was just to simplify this class and reduce the complexity. I am happy to fix this anomaly as part of a new PR. WDYT ?

Reply:
Ok, IMO creating an issue in the backlog is another way to close a problem and not forget it :)

Reply:
Raised an issue for this https://github.com/cbeust/testng/issues/1591

Reply:
👍